src/hotspot/cpu/x86/x86.ad

```diff
@@ -2415,26 +2415,113 @@ class FusedPatternMatcher {
+        })) {
+      return true;
+    }
+  } else if (n->Opcode() == Op_AddL) {
+    // Similar to AddI
+    if (VM_Version::supports_fast_2op_lea() &&
+        m->Opcode() == Op_LShiftL && m->in(2)->find_int_con(7) <= 3 &&
+        check_all_uses(m, [](Node* n){ return n->Opcode() == Op_AddL; })) {
+      return true;
+    }
+    if (VM_Version::supports_fast_3op_lea() && m->Opcode() == Op_AddL &&
+        check_all_uses(m, [](Node* n){
+          return n->Opcode() == Op_AddL && n->in(2)->Opcode() == Op_ConL &&
+                 AbstractAssembler::is_simm32(n->in(2)->get_long());
+        })) {
+      return true;
+    }
+  }
+#endif
+
+  return false;
+}
+
+// This function tries to check the bmi instruction patterns and clone the inputs
+// to preserve the matchability
 static bool is_bmi_pattern(Node* n, Node* m) {
   assert(UseBMI1Instructions, "sanity");
-  if (n != NULL && m != NULL) {
-    if (m->Opcode() == Op_LoadI) {
-      FusedPatternMatcher<TypeInt> bmii(n, m, Op_ConI);
-      return bmii.match(Op_AndI, -1, Op_SubI,  1,  0)  ||
-             bmii.match(Op_AndI, -1, Op_AddI, -1, -1)  ||
-             bmii.match(Op_XorI, -1, Op_AddI, -1, -1);
-    } else if (m->Opcode() == Op_LoadL) {
-      FusedPatternMatcher<TypeLong> bmil(n, m, Op_ConL);
-      return bmil.match(Op_AndL, -1, Op_SubL,  1,  0) ||
-             bmil.match(Op_AndL, -1, Op_AddL, -1, -1) ||
-             bmil.match(Op_XorL, -1, Op_AddL, -1, -1);
+  if (n->Opcode() == Op_AndI || n->Opcode() == Op_XorI) {
+    // Match n = ~a & b into andn n, a, b if all uses of ~a are AndI
+    if (n->Opcode() == Op_AndI && m->Opcode() == Op_XorI &&
+        m->in(2)->find_int_con(0) == -1 &&
+        check_all_uses(m, [](Node* n){ return n->Opcode() == Op_AndI; })) {
+      return true;
+    }
+    // Match n = a & (a - 1) into blsr n, a, this may reduce the liverange of a - 1
+    // Match n = a & -a into blsi n, a, this may reduce the liverange of -a
+    // Match n = a ^ (a - 1) into blsmsk n, a, this may reduce the liverange of a - 1
+
+    // If a is a load, since it appears 2 times in the match, we need to clone it
+    // so the rules with memory operands can be matched
+    if (m->Opcode() != Op_LoadI) {
+      Node* other = n->in((n->in(1) == m) ? 2 : 1);
+      m = other;
+    }
+    FusedPatternMatcher<TypeInt> bmii(n, m, Op_ConI);
+    return bmii.match(Op_AndI, -1, Op_SubI,  1,  0)  ||
+           bmii.match(Op_AndI, -1, Op_AddI, -1, -1)  ||
+           bmii.match(Op_XorI, -1, Op_AddI, -1, -1);
+  } else if (n->Opcode() == Op_AndL || n->Opcode() == Op_XorL) {
+
+#ifdef _LP64
+    // See above
+    if (n->Opcode() == Op_AndL && m->Opcode() == Op_XorL &&
+        m->in(2)->find_long_con(0) == -1 &&
+        check_all_uses(m, [](Node* n){ return n->Opcode() == Op_AndL; })) {
+      return true;
     }
+    if (m->Opcode() != Op_LoadL) {
+      Node* other = n->in((n->in(1) == m) ? 2 : 1);
+      m = other;
+    }
```

Comment:
The caller pd_clone_node is checking if m needs to be cloned, but we are changing m here locally before doing the pattern match? The original code was not doing this. Would it now result in cloning the wrong node?